guava/src/com/google/common/hash/AbstractByteHasher.java

```diff
@@ -38,19 +41,24 @@
   /** Updates this hasher with the given byte. */
   protected abstract void update(byte b);
 
+  @SuppressWarnings("upperbound:argument.type.incompatible")// `0 + b.length - 1` is < b.length.
   /** Updates this hasher with the given bytes. */
   protected void update(byte[] b) {
     update(b, 0, b.length);
   }
 
   /** Updates this hasher with {@code len} bytes starting at {@code off} in the given buffer. */
-  protected void update(byte[] b, int off, int len) {
+  protected void update(byte[] b, @NonNegative @LTLengthOf(value = "#1",offset = "#3 - 1") int off, @NonNegative @LTLengthOf(value = "#1",offset = "#2 - 1") int len) {
     for (int i = off; i < off + len; i++) {
       update(b[i]);
     }
   }
 
   /** Updates this hasher with bytes from the given buffer. */
+  @SuppressWarnings({"lowerbound:argument.type.incompatible","upperbound:argument.type.incompatible"})/*
+  Invariants: mark <= position <= limit <= capacity(is length of buffer). Since `b.remaining()` return limit - position,
```

Comment:
What is `mark`?

Reply:
More generally, the invariant here is difficult to understand. I'm not sure there's a better way to state it, because it's rather complicated, but if possible you should try. As a minor improvement, try to keep lines < 120 or so characters, so that github's diff tool (and many IDEs) doesn't render them offscreen.